### ISSUE TITLE
FBP with ParallelBeam is off when setting the voxel_size/detector spacing small enough

### ISSUE THREAD

**Issue by dveni** [carterbox/torch-radon#20]
Hi there!

First of all, thanks for keeping torchradon alive! I have been using it and I have found some issues when trying to reproduce the fbp example when the detector spacing / voxel size is smaller than 0.1. I would expect that the output should be the same always that the ratio between pixel size and voxel size is constant. However this is not true if I set `voxel_size` and `det_spacing` to 0.01 or less.

The code I'm using is adapted from the example:

```python

import os
import numpy as np
import torch
from PIL import Image
from torch_radon import ParallelBeam
from torch_radon.volumes import Volume2D
from skimage.data import shepp_logan_phantom

def to_255_numpy(x: torch.Tensor):
    xmin, xmax = x.min(), x.max()
    y = (x - xmin) / (xmax - xmin) * 255
    y = np.uint8(y.detach().cpu().numpy())
    return y


if __name__ == '__main__':
    os.environ['CUDA_VISIBLE_DEVICES'] = '0'  # set your own GPU id
    device = torch.device('cuda')
    img = shepp_logan_phantom()
    image_size = img.shape[0]
    det_count = int(image_size * 1.5)  # a det_count close to image_size will result in a circle artifact
    n_angles = 720
    print('Detector count:', det_count)
    print('Number of angles:', n_angles)

    output_dir = os.path.dirname(os.path.abspath(__file__))
    output_dir = os.path.join(output_dir, 'example_output')
    os.makedirs(output_dir, exist_ok=True)

    angles = np.linspace(0, np.pi, n_angles, endpoint=False)
    voxel_size = torch.tensor([1., 1.])*0.01
    pixel_size = 0.01
    volume = Volume2D(voxel_size=voxel_size)
    parallel_radon = ParallelBeam(det_count, angles, det_spacing=pixel_size, volume=volume)
    x = torch.FloatTensor(img).to(device)

    with torch.no_grad():
        s_par = parallel_radon.forward(x)
        print("Parallel-beam sinogram info: ", s_par.shape, s_par.min(), s_par.max())
        s_par_filt = parallel_radon.filter_sinogram(s_par)
        x_par = parallel_radon.backward(s_par_filt)
        print("Parallel-beam FBP info: ", x_par.shape, x_par.min(), x_par.max())
        print("Parallel-beam FBP error: ", ((x - x_par)**2).mean())


    Image.fromarray(to_255_numpy(x)).save(os.path.join(output_dir, 'orig.png'))
    Image.fromarray(to_255_numpy(s_par)).save(os.path.join(output_dir, 'sino_par_torchradonv2.png'))
    Image.fromarray(to_255_numpy(s_par_filt)).save(os.path.join(output_dir, 'sino_par_filt_torchradonv2.png'))
    Image.fromarray(to_255_numpy(x_par)).save(os.path.join(output_dir, 'recon_par_torchradonv2.png'))
```

When I set det_spacing and voxel_size to 0.01 the reconstruction looks like the object is shifted:

<img width="400" height="400" alt="Image" src="https://github.com/user-attachments/assets/bf9f30fd-a7a9-4f87-97b3-30cf5484a8d4" />

If I set it to 0.001 or less the reconstruction is just messed up:

<img width="400" height="400" alt="Image" src="https://github.com/user-attachments/assets/0c5d6dae-aeb5-4224-bf8a-59baca178a24" />

I must add that the sinograms look okay regardless of the spacing I'm setting, so I guess there could be a bug with the backward operator. I'm also wondering whether this is an intended use case or whether there could be something wrong with my environment.

Pointers to debug this are really appreciated!

**Comment by duchenhe** [comment]
Thanks for reporting this. I ran into the same issue and traced it to the backprojection kernel.

The forward projection is mostly fine, which matches your observation that the sinograms look correct. The problem is in `src/backward.cu`, where the backward kernels compute voxel-center physical coordinates incorrectly when `voxel_size != 1`.

The old code computes the coordinates like this:

```cpp
const float dx = (float(x) - cx) * vol_cfg.sx + vol_cfg.dx + 0.5f;
const float dy = (float(y) - cy) * vol_cfg.sy + vol_cfg.dy + 0.5f;
```

and for 3D:

```cpp
const float dz =
  (float(z) - cz) * vol_cfg.sz + vol_cfg.dz + 0.5f - proj_cfg.initial_z;
```

The issue is that the `+ 0.5f` half-voxel offset is added after converting from index coordinates to physical coordinates. That is only correct when the voxel size is exactly `1.0`.

For a general voxel size, the half-voxel offset must be applied in index space first, then scaled into physical space. The correct formula is:

```cpp
const float dx = (float(x) + 0.5f - cx) * vol_cfg.sx + vol_cfg.dx;
const float dy = (float(y) + 0.5f - cy) * vol_cfg.sy + vol_cfg.dy;
```

and for 3D:

```cpp
const float dz =
  (float(z) + 0.5f - cz) * vol_cfg.sz + vol_cfg.dz - proj_cfg.initial_z;
```

The difference between the old and corrected formula is:

```text
0.5 * (1 - voxel_size)
```

on each axis. So when `voxel_size` is small, e.g. `0.01`, the backprojection samples the sinogram at shifted physical coordinates. This explains why the FBP reconstruction appears shifted or breaks down for very small `voxel_size` / `det_spacing`, even when the ratio between them is kept constant.

This also affects autograd, because the custom backward of the forward operator uses the same backprojection kernel as the gradient.

I opened PR #21 to fix this:
https://github.com/carterbox/torch-radon/pull/21

The PR updates the voxel-center coordinate calculation in `backward.cu` and adds regression tests for non-unit voxel sizes in both 2D parallel-beam and 3D cone-beam backprojection.
